sys/src/9/amd64/sdiahci.c

```diff
@@ -818,38 +809,30 @@ countbits(uint32_t u)
 static int
 ahciconf(Ctlr *ctlr)
 {
-	Ahba *h;
-	uint32_t u;
+	Ahba *h = ctlr->hba = (Ahba*)ctlr->mmio;
 
-	h = ctlr->hba = (Ahba*)ctlr->mmio;
-	u = h->cap;
+	// Enable ahci mode
+	h->ghc |= Hae;
 
-	if((u&Hsam) == 0)
-		h->ghc |= Hae;
+	if (Intel(ctlr)) {
+		// Port control and status register
+		// Present on ICH9
+		// Not present on 100 series
+		// Enable all ports (also OOB failure retries)
+		/* configure drives 0-5 as ahci sata  (c.f. errata) */
+		pcicfgw16(ctlr->pci, 0x92, pcicfgr16(ctlr->pci, 0x92) | 0xf);
```

Comment:
I worry a bit here, given Intel's penchant for moving registers around ... should you check the devid?

Reply:
yeah...  on my controller, the register had been removed.  I'd been considering lifting the devid and some of the categorisation from the linux ahci.c code.  That can be used for this purpose.